### DISCUSSION TITLE
Session Management / Multi-User Capabilities

### DISCUSSION THREAD

**Discussion by dasmy** [ricklamers/gpt-code-ui#67] [Ideas]
With the current setup and configuration, there can apparently only be one active user at the same time.
If multiple users try to access the UI, they e.g. get results for other user's execution because no user identification or session management do exist. Even having two tabs open at the same time makes the application stop working.

Essentially, we would need the capability to have one python kernel per user, individual snakemq queues and one unique session ID per user. Then all internal messages will need to get a the session ID assigned so that they can be routed back properly.

I really have no idea how to fix this without a major re-engineering. Maybe someone else does?

**Comment by ricklamers** [comment]
It shouldn't be very difficult to create the necessary isolation, but yeah, it's quite a bit of work.

**Reply by somakaushik98** [reply]
Hi Rick,
Could you provide any mechanism where you can scale this application to mutli-users.

**Reply by ricklamers** [reply]
It’s tricky to get right, but would welcome a PR!

**Reply by dasmy** [reply]
A while ago, I started an implementation for individual sessions here: https://github.com/dasmy/gpt-code-ui/tree/dev/session_management but made several additions and changes in a non-public fork of the project. As it grew pretty organically (with several more features and fixes), publishing it and creating well-separated pull requests might be a bit of work, though...
